app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs

```diff
@@ -0,0 +1,440 @@
+                let right = left +  vis_size.x;
+                dom_entry_root.set_style_or_warn("top", format!("{}px", top));
+                dom_entry_root.set_style_or_warn("left", format!("{}px", -left));
+                grid_view::Viewport {top,bottom,left,right}
+            }));
+            text_grid.set_viewport <+ viewport;
+
+        }
+
+
+        // === Style ===
+
+        frp::extend! { network
+            init <- source::<()>();
+
+            theme_update <- all(init, font_name, font_size);
+            text_grid.set_entries_params <+  theme_update.map(f!([dom_entry_root]((_, font_name, font_size)) {
+                let parent = Some(dom_entry_root.clone_ref());
+                let font_name = font_name.clone();
+                let font_size = *font_size;
+                entry::Params { parent, font_name, font_size}
+            }));
+
+            item_width_update <- all(init, font_name, font_size);
+            item_width <- item_width_update.map(f!([]((_, font_name, font_size)) {
+                let font_size = *font_size;
+                let char_width = measure_character_width(font_name, font_size);
+                CHARS_PER_CELL * char_width
+            })).on_change();
+            item_update <- all(init, item_width, font_size);
+            text_grid.set_entries_size <+ item_update.map(f!([]((_, item_width, item_height)) {
+                Vector2::new(*item_width, *item_height)
+            }));
+
+        }
+
+        init.emit(());
+
+        self
+    }
+}
+
+impl<T> From<TextGrid<T>> for visualization::Instance {
+    fn from(t: TextGrid<T>) -> Self {
+        Self::new(&t, &t.frp, &t.network, Some(t.model.dom_entry_root.clone_ref()))
+    }
+}
+
+impl<T> display::Object for TextGrid<T> {
+    fn display_object(&self) -> &display::object::Instance {
+        &self.root
+    }
+}
+
+// ====================
+// === TextProvider ===
+// ====================
+
+/// Trait for providing text for the TextGrid.
+pub trait TextProvider {
+    /// Return a slice of the text.
+    ///
+    /// The slice is indexed by the line and the "chunk", where a chunk is a sequence of characters
+    /// of fixed length, into which the line is divided. For example, for "abcdef" there could
+    /// be chunks of size two: ["ab", "cd", "ef"], or of size three ["abc", "def"].
+    fn get_slice(&self, line: usize, chunk_size: usize, chunk_index: usize) -> Option<String>;
+
+    /// Return the number of lines.
+    fn line_count(&self) -> usize;
+
+    /// Return the length of the longest line in characters.
+    fn longest_line(&self) -> usize;
+}
+
+impl TextProvider for String {
+    fn get_slice(&self, line: usize, chunk_size: usize, chunk_index: usize) -> Option<String> {
+        self.lines().nth(line).and_then(|line| {
+            line.chars()
+                .chunks(chunk_size)
+                .into_iter()
+                .nth(chunk_index)
+                .map(|chunk| chunk.collect::<String>())
+        })
+    }
```

Comment:
This seems like a performance pitfall: Time spent scanning strings in this function will scale quadratically with the size of the input (the number of times the function is called is linear in the length of the input, and the operation itself is linear-time in the length of the input).

I can see that, due to the need to call this from FRP, we cannot simply return an iterator that would do this in linear time by maintaining a constant-space state. We could however implement the "stateless" API FRP requires in linear time at the cost of a linear-space cache: Keep a `Vec` in the model storing the end-offsets of the rows whose end-offsets have been computed; when the `str` for a row is needed, if it is beyond the end of the cache, start scanning lines from the last known offset; extend the cache with any newly-computed values. (It would be simpler if we could be sure rows will be requested in order, but I assume we can't rely on that.)

Reply:
This is only a dummy API for the demo scene. The actual data will be provided by the engine and will follow a different pattern that has a local cache. This will be added in the follow-up PR. But a similar issue might exist on the Enso/Engine side, where we do need to slice the full string/data.